File: backends/java/zappy-domain/src/main/java/com/zappymart/domain/ordering/Order.java
package com.zappymart.domain.ordering;

import com.zappymart.domain.cart.Cart;
import com.zappymart.domain.cart.CartLine;
import com.zappymart.domain.shared.Money;
import com.zappymart.domain.shared.Result;
import com.zappymart.domain.shared.UserErrorCode;

import java.time.Instant;
import java.util.List;
import java.util.Objects;

public record Order(
        String id,
        String number,
        String customerId,
        OrderStatus status,
        List<OrderLine> lines,
        String promotionCode,
        Money subtotal,
        Money discount,
        Money shipping,
        Money total,
        Instant placedAt) {

    public Order {
        Objects.requireNonNull(id, "id");
        Objects.requireNonNull(number, "number");
        Objects.requireNonNull(customerId, "customerId");
        Objects.requireNonNull(status, "status");
        lines = List.copyOf(lines);
        Objects.requireNonNull(subtotal, "subtotal");
        Objects.requireNonNull(discount, "discount");
        Objects.requireNonNull(shipping, "shipping");
        Objects.requireNonNull(total, "total");
        Objects.requireNonNull(placedAt, "placedAt");
        if (lines.isEmpty()) {
            throw new IllegalArgumentException("An order holds at least one line");
        }
    }

    public static Result<Order> place(String id, String number, String customerId, Cart cart, Instant moment) {
        if (!cart.hasLines()) {
            return Result.refuse(UserErrorCode.CART_EMPTY, "The cart has no lines, so there is nothing to order.");
        }
        for (CartLine line : cart.lines()) {
            if (!line.product().hasStockFor(line.quantity())) {
                return Result.refuse(UserErrorCode.OUT_OF_STOCK,
                        "There is not enough stock of " + line.product().name() + ", so no order was placed.");
            }
        }
        List<OrderLine> orderLines = cart.lines().stream()
                .map(line -> new OrderLine(line.product().id(), line.product().name(),
                        line.product().price(), line.quantity()))
                .toList();
        String appliedCode = cart.promotionCode() == null ? null : cart.promotionCode().value();
        return Result.of(new Order(id, number, customerId, OrderStatus.PAID, orderLines, appliedCode,
                cart.subtotal(), cart.discount(), cart.shipping(), cart.total(), moment));
    }

    public OrderPlaced placementEvent() {
        return new OrderPlaced(id, number, customerId, promotionCode, placedAt);
    }
}

File: backends/kotlin/zappy-domain/src/main/kotlin/nl/zappymart/domain/ordering/Order.kt
package nl.zappymart.domain.ordering

import java.time.Instant
import nl.zappymart.domain.cart.Cart
import nl.zappymart.domain.shared.Money
import nl.zappymart.domain.shared.PromotionCode
import nl.zappymart.domain.shared.Result
import nl.zappymart.domain.shared.UserErrorCode
import nl.zappymart.domain.shared.refusal

data class Order(
    val id: String,
    val number: String,
    val customerId: String,
    val status: OrderStatus,
    val lines: List<OrderLine>,
    val promotionCode: PromotionCode?,
    val subtotal: Money,
    val discount: Money,
    val shipping: Money,
    val total: Money,
    val placedAt: Instant,
) {

    fun placed() = OrderPlaced(id, number, customerId, promotionCode, placedAt)

    companion object {

        fun place(cart: Cart, customerId: String, id: String, number: String, moment: Instant): Result<Order> {
            if (cart.isEmpty) {
                return refusal(UserErrorCode.CART_EMPTY, "There is nothing in the cart to order.")
            }
            val short = cart.lines.firstOrNull { line -> !line.product.hasStockFor(line.quantity) }
            if (short != null) {
                return refusal(
                    UserErrorCode.OUT_OF_STOCK,
                    "${short.product.name} has ${short.product.stock} in stock and ${short.quantity} were ordered.",
                )
            }
            return Result.Success(
                Order(
                    id = id,
                    number = number,
                    customerId = customerId,
                    status = OrderStatus.PAID,
                    lines = cart.lines.map { line ->
                        OrderLine(line.product.id, line.product.name, line.product.price, line.quantity)
                    },
                    promotionCode = cart.promotion?.code,
                    subtotal = cart.subtotal,
                    discount = cart.discount,
                    shipping = cart.shipping,
                    total = cart.total,
                    placedAt = moment,
                ),
            )
        }
    }
}
